core/ldap_api.php

```diff
@@ -207,76 +189,107 @@ function ldap_escape_string( $p_string ) {
- * @return string The field value or null if not found.
+ * @param string $p_username The username.
+ *
+ * @return array|false User data, false if not found or errors occurred
  */
-function ldap_get_field_from_username( $p_username, $p_field ) {
-	$t_ldap_organization    = config_get( 'ldap_organization' );
-	$t_ldap_root_dn         = config_get( 'ldap_root_dn' );
-	$t_ldap_uid_field		= config_get( 'ldap_uid_field' );
+function ldap_cache_user_data( $p_username ) {
+	global $g_cache_ldap_data;
 
-	$c_username = ldap_escape_string( $p_username );
+	# Returne cached data if available
+	if( isset( $g_cache_ldap_data[$p_username] ) ) {
+		return $g_cache_ldap_data[$p_username];
+	}
 
-	log_event( LOG_LDAP, 'Retrieving field \'' . $p_field . '\' for \'' . $p_username . '\'' );
+	log_event( LOG_LDAP, "Retrieving data for '$p_username' from LDAP server" );
 
 	# Bind
 	log_event( LOG_LDAP, 'Binding to LDAP server' );
 	$t_ds = @ldap_connect_bind();
 	if( $t_ds === false ) {
 		ldap_log_error( $t_ds );
-		return null;
+		return false;
 	}
 
 	# Search
-	$t_search_filter        = '(&' . $t_ldap_organization . '(' . $t_ldap_uid_field . '=' . $c_username . '))';
-	$t_search_attrs         = array( $t_ldap_uid_field, $p_field, 'dn' );
+	$t_ldap_organization = config_get( 'ldap_organization' );
+	$t_ldap_root_dn      = config_get( 'ldap_root_dn' );
+	$t_ldap_uid_field    = config_get( 'ldap_uid_field' );
+
+	$t_search_filter = '(&' . $t_ldap_organization
+		. '(' . $t_ldap_uid_field . '=' . ldap_escape_string( $p_username ) . '))';
+	$t_search_attrs = array(
+		'mail',
+		config_get( 'ldap_realname_field' )
+	);
 
 	log_event( LOG_LDAP, 'Searching for ' . $t_search_filter );
 	$t_sr = @ldap_search( $t_ds, $t_ldap_root_dn, $t_search_filter, $t_search_attrs );
 	if( $t_sr === false ) {
 		ldap_log_error( $t_ds );
 		ldap_unbind( $t_ds );
 		log_event( LOG_LDAP, 'ldap search failed' );
-		return null;
+		return false;
 	}
 
 	# Get results
-	$t_info = ldap_get_entries( $t_ds, $t_sr );
-	if( $t_info === false ) {
-		ldap_log_error( $t_ds );
-		log_event( LOG_LDAP, 'ldap_get_entries() returned false.' );
-		return null;
+	$t_entry = ldap_first_entry( $t_ds, $t_sr );
+	if( $t_entry === false ) {
+		log_event( LOG_LDAP, 'No matches found.' );
+		$g_cache_ldap_data[$p_username] = false;
+		return false;
 	}
 
+	$t_data = false;
+	foreach( $t_search_attrs as $t_attr ) {
+		# Suppress error to avoid Warning in case an invalid attribute was specified
+		$t_value = @ldap_get_values( $t_ds, $t_entry, $t_attr );
+		if( $t_value === false ) {
+			log_event( LOG_LDAP, "WARNING: field '$t_attr' does not exist" );
+			continue;
+		}
+		$t_data[$t_attr] = $t_value[0];
+	}
+
+	# Store data in the cache
+	$g_cache_ldap_data[$p_username] = $t_data;
+
 	# Free results / unbind
+	# According to documentation, calling ldap_free_result() is not strictly necessary
 	log_event( LOG_LDAP, 'Unbinding from LDAP server' );
-	ldap_free_result( $t_sr );
+	# ldap_free_result( $t_sr );
```

Comment:
Why leave the code commented out?  Either do it anyways (which I would do if we expect something to be called many times in one page) or remove the commented code.  `ldap_free_result()` is also called in other places in `ldap_api.php`.

Reply:
I initially commented the statement out for testing purposes, and also because I was considering a more in-depth refactoring.

The idea was to keep the LDAP connection opened throughout script execution in order to reduce overhead and further improve performance, but I decided not to implement it for now and focus on fixing the immediate performance problem.

I might still do this in a later stage if time allows.

> Either do it anyways (which I would do if we expect something to be called many times in one page

The [documentation](https://www.php.net/manual/en/function.ldap-free-result.php) says:
_In case the script is making successive searches which return large result sets, ldap_free_result() could be called to keep the runtime memory usage by the script low_

We do make multiple ldap searches on a page, but they should normally only return one entry each, so I don't think that qualifies as _large result sets_.

> ldap_free_result() is also called in other places in ldap_api.php.

The other calls are in ldap_authenticate_by_username() which in usual scenarios is only done once. But I guess they should be removed as well for consistency.